### PR TITLE
Add client notifications from discovery

### PR DESCRIPTION
**Pull request by jowlee**
### Description
Client Reads for notifications from discovery

### Dragons
Behind a feature flag, but changes how notifications are read.

### How Has This Been Tested?
Ran against a local discovery box pointed at a prod discovery database and validated that each of the notification types were rendered

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?
notifications panel on web 

### Feature Flags ###
Yes, new discovery notifications read feature flag

### CODE REVIEW

File: packages/web/src/common/store/notifications/sagas.ts [hunk 1]
```diff
@@ -547,6 +574,9 @@ export function* getNotifications(isFirstFetch: boolean) {
             withDethroned
           })
         )
+      if (notificationsResponse && 'notifications' in notificationsResponse) {
+        notificationsResponse.notifications = notifications
+      }
```
Comment (jowlee) [inline packages/web/src/common/store/notifications/sagas.ts:579]: Note there is a follow up to get the unread count and playlist update notifications from discovery as well so we won't have to rely on the initial identity call soon

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/jowlee-client-read-notif

File: packages/common/src/services/audius-backend/AudiusBackend.ts [hunk 2]
```diff
@@ -2069,6 +2073,303 @@ export const audiusBackend = ({
...
+    return {
+      groupId: notification.group_id,
+      timestamp: new Date(timestamp * 1000).toString(),
+      isViewed: notification.seen_at == null,
+      id: `timestamp:${timestamp}:group_id:${notification.group_id}`
+    }
+  }
+
+  function mapDiscoveryNotification(notification: DiscoveryNotification) {
+    console.log({ notification })
```
Comment (dylanjeffers) [inline packages/common/src/services/audius-backend/AudiusBackend.ts:2094]: log

File: packages/common/src/services/audius-backend/AudiusBackend.ts [hunk 3]
```diff
@@ -2069,6 +2073,303 @@ export const audiusBackend = ({
...
+        type: NotificationType.Follow,
+        userIds,
+        ...formatBaseNotification(notification)
+      }
+    } else if (notification.type === 'repost') {
+      let entityId
+      let entityType
+      const userIds = notification.actions.map((action) => {
+        const data = action.data
+        entityId = decodeHashId(data.repost_item_id)
```
Comment (dylanjeffers) [inline packages/common/src/services/audius-backend/AudiusBackend.ts:2110]: im a little confused with the let + map for entityId and entitiyType, does it just become the values of the last action?

File: packages/common/src/services/audius-backend/AudiusBackend.ts [hunk 4]
```diff
@@ -2069,6 +2073,303 @@ export const audiusBackend = ({
...
+    const timestamp = notification.actions[0].timestamp
+    return {
+      groupId: notification.group_id,
+      timestamp: new Date(timestamp * 1000).toString(),
+      isViewed: notification.seen_at == null,
+      id: `timestamp:${timestamp}:group_id:${notification.group_id}`
+    }
+  }
+
+  function mapDiscoveryNotification(notification: DiscoveryNotification) {
```
Comment (dylanjeffers) [inline packages/common/src/services/audius-backend/AudiusBackend.ts:2093]: this all totally makes sense, but long term are you thinking we move our data types to be in line with discovery-notification data?

File: packages/common/src/services/audius-backend/AudiusBackend.ts [hunk 5]
```diff
@@ -2093,7 +2394,7 @@ export const audiusBackend = ({
         : ''
       // TODO: withRemix, withTrending, withRewards are always true and should be removed in a future release
       const notifications = await fetch(
-        `${identityServiceUrl}/notifications?${limitQuery}${timeOffsetQuery}${handleQuery}${withDethronedQuery}&withTips=true&withRewards=true&withRemix=true&withTrendingTrack=true`,
+        `${identityServiceUrl}/notifications?${limitQuery}${timeOffsetQuery}${handleQuery}${withDethronedQuery}}&withTips=true&withRewards=true&withRemix=true&withTrendingTrack=true`,
```
Comment (dylanjeffers) [inline packages/common/src/services/audius-backend/AudiusBackend.ts:2397]: is there a libs method for this btw, or we just doing pure manual for this? also do we need withRemix, withTrending, withRewards needed these days?

File: packages/common/src/store/notifications/selectors.ts [hunk 6]
```diff
@@ -86,8 +86,9 @@ export const getNotificationUser = (
   notification: Notification
 ) => {
   if (
-    notification.type === NotificationType.Milestone &&
-    notification.achievement === Achievement.Followers
+    notification.type === NotificationType.TierChange ||
```
Comment (dylanjeffers) [inline packages/common/src/store/notifications/selectors.ts:89]: nice add, so was this just missing in general?

File: packages/common/src/store/notifications/types.ts [hunk 7]
```diff
@@ -104,8 +104,230 @@ export type BaseNotification = {
...
+}
+
+export type DiscoveryAction<T> = {
+  specifier: string
+  type: string
+  timestamp: number
+  data: T
+}
+
+export type DiscoveryAnnouncementNotificationAction = {
```
Comment (dylanjeffers) [inline packages/common/src/store/notifications/types.ts:125]: sweet thanks for adding all these types, def curious if these can become the base of the new notification types?

File: packages/web/src/common/store/notifications/sagas.ts [hunk 8]
```diff
@@ -145,6 +145,11 @@ export function* fetchNotifications(action: FetchNotifications) {
         getFeatureEnabled,
         FeatureFlags.SUPPORTER_DETHRONED_ENABLED
       )) as boolean | null) ?? false
+    const useDiscoveryNotifications =
```
Comment (dylanjeffers) [inline packages/web/src/common/store/notifications/sagas.ts:148]: def a tragic situation with the double (( and casting... will make a note to improve those types

File: packages/web/src/common/store/notifications/sagas.ts [hunk 9]
```diff
@@ -153,6 +158,22 @@ export function* fetchNotifications(action: FetchNotifications) {
         withDethroned
       })
     )
+
+    if (useDiscoveryNotifications && 'notifications' in notificationsResponse) {
+      const discoveryNotifications = yield* call(() => {
```
Comment (dylanjeffers) [inline packages/web/src/common/store/notifications/sagas.ts:163]: kinda a nit, but i think you can just create all the variables outside the call, and then just do `call(audiusBackendInstance.getDiscoveryNotifications, ...)`

File: packages/web/src/common/store/notifications/sagas.ts [hunk 10]
```diff
@@ -547,6 +573,18 @@ export function* getNotifications(isFirstFetch: boolean) {
             withDethroned
           })
         )
+
+      if (
+        useDiscoveryNotifications &&
+        notificationsResponse &&
+        'notifications' in notificationsResponse
+      ) {
+        const discoveryNotifications = yield* call(() => {
```
Comment (dylanjeffers) [inline packages/web/src/common/store/notifications/sagas.ts:582]: `call(audiusbackendInstance.getDiscovery, {})`

File: packages/web/src/components/notification/NotificationPanel.tsx [hunk 11]
```diff
@@ -146,16 +146,14 @@ export const NotificationPanel = ({ anchorRef }: NotificationPanelProps) => {
                 getScrollParent={getScrollParent}
               >
                 <div className={styles.content}>
-                  {notifications
-                    .filter(({ isHidden }: any) => !isHidden)
```
Comment (dylanjeffers) [inline packages/web/src/components/notification/NotificationPanel.tsx:150]: that's cool so was this just never used or handled somewhere else?

**Review by dylanjeffers — APPROVED** [review]
awesome work here! excited to try this out

File: packages/web/src/components/notification/NotificationPanel.tsx (hunk 11 above)
Comment (jowlee) [reply packages/web/src/components/notification/NotificationPanel.tsx:150]: this feature was removed

File: packages/web/src/common/store/notifications/sagas.ts (hunk 10 above)
Comment (jowlee) [reply packages/web/src/common/store/notifications/sagas.ts:582]: yeah, nice - will change

File: packages/web/src/common/store/notifications/sagas.ts (hunk 8 above)
Comment (jowlee) [reply packages/web/src/common/store/notifications/sagas.ts:148]: yeah, unf and thx!

File: packages/common/src/store/notifications/selectors.ts (hunk 6 above)
Comment (jowlee) [reply packages/common/src/store/notifications/selectors.ts:89]: well we were filling the field on another part of the notification I believe

File: packages/common/src/services/audius-backend/AudiusBackend.ts (hunk 5 above)
Comment (jowlee) [reply packages/common/src/services/audius-backend/AudiusBackend.ts:2397]: just manually - the whole idea before was that notifications was audius client specific and not not external consumption - so we never added to libs

File: packages/common/src/services/audius-backend/AudiusBackend.ts (hunk 4 above)
Comment (jowlee) [reply packages/common/src/services/audius-backend/AudiusBackend.ts:2093]: yea fo sho
was just mapping the new notifs to the old as the easiest way to get this done without potentially breaking stuff

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/jowlee-client-read-notif

**Pull request merged**